src/python/pants/engine/target_test.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from collections import OrderedDict
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import ClassVar, List, Optional, Tuple
+
+import pytest
+
+from pants.engine.fs import EMPTY_DIRECTORY_DIGEST, PathGlobs, Snapshot
+from pants.engine.rules import UnionMembership, rule
+from pants.engine.selectors import Get
+from pants.engine.target import AsyncField, PrimitiveField, Target
+from pants.testutil.engine.util import MockGet, run_rule
+from pants.util.collections import ensure_str_list
+from pants.util.memo import memoized_property
+from pants.util.ordered_set import OrderedSet
+
+
+class HaskellGhcExtensions(PrimitiveField):
+    alias: ClassVar = "ghc_extensions"
+    raw_value: Optional[List[str]]
+
+    @memoized_property
+    def value(self) -> List[str]:
+        if self.raw_value is None:
+            return []
+        # Add some arbitrary validation to test that hydration works properly.
+        bad_extensions = [
+            extension for extension in self.raw_value if not extension.startswith("Ghc")
+        ]
+        if bad_extensions:
+            raise ValueError(
+                f"All elements of `{self.alias}` must be prefixed by `Ghc`. Received "
+                f"{bad_extensions}."
+            )
+        return self.raw_value
+
+
+class HaskellSources(AsyncField):
+    alias: ClassVar = "sources"
+    raw_value: Optional[List[str]]
+
+    def validate_pre_hydration(self) -> None:
```

Comment:
So, `validate_pre_hydration` and `validate_post_hydration` feel like they can/should go in the `@rule` that hydrates the field, otherwise your field definition API requires extending three different methods.

Was the idea here that someone might want to change the rule that hydrates a field, while keeping the validation of the field the same? That feels like a pattern that we could wait to see happening in the wild before adding support for... or one that we should explicitly label here as something that we're anticipating (and should remove complexity for if we don't see the pattern in practice.)

Reply:
As discussed over video chat, there is no super compelling reason to have these methods, other than the principle of proximity and trying to give some structure to `AsyncField`. I'll remove them for now to avoid premature optimization.